### PR TITLE
Problem: Change Coin to (de)serailize from and to string(CRO-125)

### PR DESCRIPTION
**Pull request by leejw51crypto**
Solution:  refactoring api call

### CODE REVIEW

File: chain-core/src/init/coin.rs [hunk 1]
```diff
@@ -5,23 +5,22 @@
...
 use crate::state::tendermint::TendermintVotePower;
+use crate::state::tendermint::TENDERMINT_MAX_VOTE_POWER;
 use parity_codec::{Decode, Encode, Input};
 
-use crate::state::tendermint::TENDERMINT_MAX_VOTE_POWER;
-#[cfg(feature = "serde")]
-use serde::de::{Deserialize, Deserializer, Error, Visitor};
-#[cfg(feature = "serde")]
-use serde::Serialize;
+use serde::{Deserialize, Serialize, Serializer};
```
Comment (tomtau) [inline chain-core/src/init/coin.rs]: please keep the feature guards for serde:
```
#[cfg(feature = "serde")]
```

File: chain-core/src/init/coin.rs [hunk 2]
```diff
@@ -5,23 +5,22 @@
 
 use crate::init::{MAX_COIN, MAX_COIN_DECIMALS, MAX_COIN_UNITS};
 use crate::state::tendermint::TendermintVotePower;
+use crate::state::tendermint::TENDERMINT_MAX_VOTE_POWER;
 use parity_codec::{Decode, Encode, Input};
 
-use crate::state::tendermint::TENDERMINT_MAX_VOTE_POWER;
-#[cfg(feature = "serde")]
-use serde::de::{Deserialize, Deserializer, Error, Visitor};
-#[cfg(feature = "serde")]
```
Comment (devashishdxt) [inline chain-core/src/init/coin.rs]: Any particular reason for removing `serde` feature flag?

File: chain-core/src/init/coin.rs [hunk 3]
```diff
@@ -5,23 +5,22 @@
...
 use std::convert::TryFrom;
 use std::{fmt, mem, ops, result, slice};
 
 #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Encode)]
-#[cfg_attr(feature = "serde", derive(Serialize))]
 pub struct Coin(u64);
 
 /// error type relating to `Coin` operations
-#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy, serde::Serialize, Deserialize)]
```
Comment (devashishdxt) [inline chain-core/src/init/coin.rs]: Why do we have to implement `Serialize` and `Deserialize` for `CoinError`?
```suggestion
#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
```

File: chain-core/src/init/coin.rs (hunk 1 above)
Comment (leejw51crypto) [reply chain-core/src/init/coin.rs]: ok
i also add unit test

File: chain-core/src/init/coin.rs (hunk 2 above)
Comment (leejw51crypto) [reply chain-core/src/init/coin.rs]: no , just a mistake

Comment (leejw51crypto) [reply chain-core/src/init/coin.rs]: i'll restore back.

**Comment by leejw51crypto** [comment]
add unit test and restored feature guards

File: chain-core/src/init/coin.rs (hunk 3 above)
Comment (leejw51crypto) [reply chain-core/src/init/coin.rs]: if not, compiling fails

File: chain-core/src/init/coin.rs (hunk 1 above)
Comment (leejw51crypto) [reply chain-core/src/init/coin.rs]: done

File: chain-core/src/init/coin.rs (hunk 3 above)
Comment (devashishdxt) [reply chain-core/src/init/coin.rs]: Strange. I don't see any reason for compilation to fail.

Comment (leejw51crypto) [reply chain-core/src/init/coin.rs]: yes, it compiles fine. i'll remove it.

File: Cargo.lock [hunk 4]
```diff
@@ -257,7 +257,7 @@ dependencies = [
  "criterion 0.2.11 (registry+https://github.com/rust-lang/crates.io-index)",
  "digest 0.8.1 (registry+https://github.com/rust-lang/crates.io-index)",
  "env_logger 0.6.2 (registry+https://github.com/rust-lang/crates.io-index)",
- "ethbloom 0.6.4 (registry+https://github.com/rust-lang/crates.io-index)",
+ "ethbloom 0.5.3 (registry+https://github.com/rust-lang/crates.io-index)",
```
Comment (tomtau) [inline Cargo.lock]: it seems this reverts the current master -- e.g. https://github.com/crypto-com/chain/pull/178

File: client-index/src/index/default_index.rs [hunk 5]
```diff
@@ -469,15 +469,6 @@ mod tests {
         fn broadcast_transaction(&self, _: &[u8]) -> Result<()> {
             Ok(())
         }
-
-        /// Get abci query
-        fn query(&self, _path: &str, _data: &str) -> Result<QueryResult> {
```
Comment (tomtau) [inline client-index/src/index/default_index.rs]: and deleting your previous PR?

Comment (leejw51crypto) [reply client-index/src/index/default_index.rs]: oh it not merged with latest.
I'll recheck

**Comment by leejw51crypto** [comment]
only coin changed

File: chain-core/src/init/coin.rs [hunk 6]
```diff
@@ -33,6 +33,46 @@ pub enum CoinError {
     Negative,
 }
 
+impl Serialize for Coin {
```
Comment (tomtau) [inline chain-core/src/init/coin.rs]: should be feature-guarded (as imports are) `#[cfg(feature = "serde")]`

File: chain-core/src/init/coin.rs [hunk 7]
```diff
@@ -33,6 +33,46 @@ pub enum CoinError {
...
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let coin_string = self.0.to_string();
+        serializer.serialize_str(&coin_string[..])
+    }
+}
+
+impl<'de> Deserialize<'de> for Coin {
```
Comment (tomtau) [inline chain-core/src/init/coin.rs]: should be feature-guarded (as imports are) `#[cfg(feature = "serde")]`

File: Cargo.lock (hunk 4 above)
Comment (leejw51crypto) [reply Cargo.lock]: i modified and merged with latest

**Comment by devashishdxt** [comment]
> the current changes undo some of the latest master?

Here's a simple solution for dealing with this:

- Configure `upstream` in your local `chain` repository:
```
git remote add upstream git@github.com:crypto-com/chain.git
```
- Add following scripts in your `~/.gitbin`
  - [git-sync](https://gist.github.com/devashishdxt/2b1c38d81ec1ca2180ad8f9045554ab6)
  - [git-cleanup](https://gist.github.com/devashishdxt/cd85210e6ff05ae95eaf94fc9c56de17)

How this works:
- Run `git sync`: This will sync your fork's master branch with upstream master (And also your local repo).
- Run `git checkout -b feature-branch`: Create a feature branch.
- Make and commit your changes.
- Run `git sync` before pushing.
  - If there are any changes, run `git checkout feature-branch && git merge master`
- Run `git push`
- After PR is closed, run `git cleanup`. This will delete all your feature branches.

NOTE: This will only work if you don't make any manual commit on your fork's master branch. Always create a new branch to make any changes.

**Pull request merged**